### DISCUSSION TITLE
how do I use sidebars/toolboxes for input?

### DISCUSSION THREAD

**Discussion by lostmygithubaccount** [quarto-dev/quarto-cli#8574] [Q&A]
### Description

The examples here are all empty: https://quarto.org/docs/dashboards/inputs.html

I just need a few text/single select inputs for a dashboard I'm using. see code below

**Comment by lostmygithubaccount** [comment]
found an example here I believe: https://github.com/quarto-dev/quarto-web/blob/8dac33b1e847a06215d07f5a0b31b52052672cd3/docs/dashboards/_inputs.qmd#L71C1-L76

**Reply by lostmygithubaccount** [reply]
now I'm struggling to get the values out for use downstream though

**Reply by lostmygithubaccount** [reply]
somewhere like this but still not working:

````
## PyPI stats {height=100%}

### {.toolbar}

```{python}
# | panel: input

from shiny import ui, render
from datetime import datetime

test = "hello"
ui.input_text("package", "Package:", value="ibis-framework")
#ui.input_select("x", "Variable:", choices=["bill_length_mm", "bill_depth_mm"])
#ui.input_select("dist", "Distribution:", choices=["hist", "kde"])
#ui.input_checkbox("rug", "Show rug marks", value=False)
ui.input_date_range(
    "daterange",
    "Date range",
    start="2020-01-01",
    end=datetime.today().strftime("%Y-%m-%d"),
)
```

### plot

```{python}
@render.plot
def bar():
    t = (
        pypi.filter(ibis._.project == input.package())
        .group_by(date=ibis._.date.truncate("M"), version="version")
        .agg(count=ibis._.count())
        .order_by(ibis._.date.desc())
        .mutate(date=ibis._.date.cast("string"))
    )

    px.bar(
        t,
        x="date",
        y="count",
        title=f"Downloads for {package}",
        color=group_by,
    )
```
````

**Reply by mcanouil** [reply]
Could you properly format your code by using four backticks to enclose code containing code cells?

**Reply by mcanouil** [reply]
If you are looking for Shiny for Python, the best place is: https://shiny.posit.co/py/index.html

And more generally look at the interactivity for dashboard pages: https://quarto.org/docs/dashboards/interactivity/index.html

**Reply by lostmygithubaccount** [reply]
updated -- am I still doing something wrong here? the plot is empty (I'm also very confused about the magical `input` variable but that's fine):

````
## PyPI stats {height=100%}

### {.toolbar}

```{python}
# | panel: input

ui.input_text("package", "Package:", value="ibis-framework")

#ui.input_date_range(
#    "daterange",
#    "Date range",
#    start="2020-01-01",
#    end=datetime.today().strftime("%Y-%m-%d"),
#)
```

### plot

```{python}
group_by = "version"
```

```{python}
@render.plot
def pypi_plot():
    t = (
        pypi.filter(ibis._.project == input.package())
        .group_by(date=ibis._.date.truncate("M"), version="version")
        .agg(count=ibis._.count())
        .order_by(ibis._.date.desc())
        .mutate(date=ibis._.date.cast("string"))
    )

    px.bar(
        t,
        x="date",
        y="count",
        title=f"Downloads for {package}",
        color=group_by,
    )
```

````

<img width="1683" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/54814569/736c94d2-babb-42c9-b257-6ec45a4757b1">

**Reply by mcanouil** [reply]
Your code is not reproducible, so can't really help.

The option symbol should have no space in it: `#|` not `# |`. Syntax matters.

You should start playing with the example, then tweak the example, and finally try your own thing.
There are no shortcuts to learn something.

**Reply by lostmygithubaccount** [reply]
I was hoping this was something simple -- I've used Ibis, Quarto (including dashboards), and Plotly before but the documentation for how to use inputs in Quarto dashboards (linked at the top) just shows empty code. So I've been looking around for an example. The one I found I followed, but it results in empty output where the plot should be with no information to help debug (hence asking for help here). Here's a reproducible example:

```bash
pip install 'ibis-framework[duckdb,examples]' shiny itables plotly
```

(might be missing some installs)

Dashboard qmd:

````
--- 
title: "repro"
format:
  dashboard:
    theme: darkly
    scrolling: true
--- 

```{python}
#| echo: false
#| output: false
import ibis

import plotly.express as px

from shiny import ui, render
from itables import show

# set plotly express to dark mode
px.defaults.template = "plotly_dark"

ibis.options.interactive = True

t = ibis.examples.penguins.fetch()
```

## {height=25%}

```{python}
show(t.limit(10).to_pandas())
```

```{python}
show(
  t.group_by(ibis._.species)
  .agg(
    count=ibis._.count()
  )
  .order_by(ibis._["count"].desc())
  .to_pandas()
)
```

## {height=10%}

### {.toolbar}

```{python}
#| panel: input
ui.input_text("species", "Species:", value="Gentoo")
```

### plot

```{python}
@render.plot
def bar_chart():
    px.bar(
        t.filter(t.species == input.species())
        .group_by(ibis._.island)
        .agg(
            count=ibis._.count()
        ),
        x="island",
        y="count",
    )
```

````

dashboard preview:

<img width="1665" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/54814569/b3c87c87-7e5d-4b7c-9499-ff67b249147b">

**Reply by mcanouil** [reply]
Look at the links I shared: https://quarto.org/docs/dashboards/interactivity/shiny-python/index.html

**Reply by lostmygithubaccount** [reply]
I did. The only noticable difference seems to be the `server: shiny` which does not change anything when added.

Using that example there, it seems like plotly simply doesn't work (or I'm still doing something wrong):

**note:** plotly histogram code below is off but the point is plotly plots don't seem to work unless I'm missing something

````
---
title: "Penguin Bills"
format: dashboard
server: shiny
---

```{python}
import seaborn as sns
import plotly.express as px
penguins = sns.load_dataset("penguins")
```

## {.sidebar}

```{python}
from shiny import render, ui
ui.input_select("x", "Variable:",
                choices=["bill_length_mm", "bill_depth_mm"])
ui.input_select("dist", "Distribution:", choices=["hist", "kde"])
ui.input_checkbox("rug", "Show rug marks", value = False)
```

## Column

```{python}
@render.plot
def displot():
    #sns.displot(
    #    data=penguins, hue="species", multiple="stack",
    #    x=input.x(), rug=input.rug(), kind=input.dist())

    px.histogram(penguins, x=input.x(), color="species", marginal="rug",
                 histfunc="count", barmode="overlay")
```
````

<img width="1654" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/54814569/e5967f67-e15b-441c-b6af-25f072de7c39">

versus w/ the seaborn code:

<img width="1668" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/54814569/65ef9ec2-74e5-4541-ab6e-3bfad0612d31">

it's very difficult to debug with no documentation and silent errors (hence coming here to ask for help)

**Reply by mcanouil** [reply]
See <https://quarto.org/docs/troubleshooting/index.html>.
Here your issue is mostly about Python rather than Quarto.

Take a look at other discussions involving plotly:
- https://github.com/quarto-dev/quarto-cli/discussions/5264
- https://github.com/quarto-dev/quarto-cli/discussions?discussions_q=plotly

**Comment by mcanouil** [comment]
FWIW, there are two examples in the gallery: https://quarto.org/docs/gallery/#dashboards 

- https://github.com/wch/retirement-simulation-dashboard
- https://github.com/jjallaire/penguins-dashboard

And: https://shiny.posit.co/py/docs/shinylive.html

As of today, the only thing that does not seem to support `# |` in Python-like code cell is the Quarto-shinylive extension.

````qmd
---
title: "Quarto Playground"
format: html
filters:
  - shinylive
---

```{shinylive-python}
# | standalone: true

from shiny import *

app_ui = ui.page_fluid(
    ui.input_slider("n", "N", 0, 100, 40),
    ui.output_text_verbatim("txt"),
)


def server(input, output, session):
    @output
    @render.text
    def txt():
        return f"The value of n*2 is {input.n() * 2}"


app = App(app_ui, server)
```
````

**Reply by mcanouil** [reply]
- https://github.com/quarto-ext/shinylive/issues/67
